pkg/kube/pod_controller.go

```diff
@@ -230,7 +230,66 @@ func (p *podController) StreamPodLogs(ctx context.Context) (io.ReadCloser, error
 		return nil, ErrPodControllerPodNotStarted
 	}
 
-	return StreamPodLogs(ctx, p.cli, p.pod.Namespace, p.pod.Name, ContainerNameFromPodOptsOrDefault(p.podOptions))
+	return newRestoreLogStreamReader(ctx, p.cli, p.pod.Namespace, p.pod.Name, ContainerNameFromPodOptsOrDefault(p.podOptions))
+}
+
+// newRestoreLogStreamReader creates a new restoreLogStreamReader instance which is used to stream logs from a pod.
+// restoreLogStreamReader will automatically try to establish a new log stream if the current one is closed and the pod is still alive.
+// This wrapper has to exist as there is hardcoded 4h timeout in kubelet for streaming logs.
+func newRestoreLogStreamReader(ctx context.Context, cli kubernetes.Interface, namespace, podName, containerName string) (io.ReadCloser, error) {
+	reader, err := StreamPodLogs(ctx, cli, namespace, podName, containerName, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &restoreLogStreamReader{
+		ctx:           ctx,
+		cli:           cli,
+		namespace:     namespace,
+		podName:       podName,
+		containerName: containerName,
+		reader:        reader,
+	}, nil
+}
+
+type restoreLogStreamReader struct {
+	ctx           context.Context
+	cli           kubernetes.Interface
+	namespace     string
+	podName       string
+	containerName string
+	reader        io.ReadCloser
+	lastReadTime  metav1.Time
+}
+
+func (s *restoreLogStreamReader) Read(p []byte) (n int, err error) {
+	n, err = s.reader.Read(p)
+	defer func() { s.lastReadTime = metav1.Now() }()
+	if err == io.EOF {
+		pod, err := s.cli.CoreV1().Pods(s.namespace).Get(s.ctx, s.podName, metav1.GetOptions{})
+		if err != nil {
+			return 0, err
+		}
+
+		if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+			return n, io.EOF
+		}
+
+		err = s.reader.Close()
+		if err != nil {
+			return 0, err
+		}
+		s.reader, err = StreamPodLogs(s.ctx, s.cli, s.namespace, s.podName, s.containerName, &s.lastReadTime)
+		if err != nil {
+			return 0, err
+		}
+
+		return s.reader.Read(p)
```

Comment:
This looks like a stream read is only repeated once. Should we assume that it can happen multiple times and use a loop instead? I don't really see a reason why it wouldn't happen multiple times, am I missing something?

Reply:
It's very unlikely that the new stream will return io.EOF, and even if it does, we should not assume it is for the same reason that we try to fix in this PR. A safer option will be to resume the connection only once after a successful read.

Reply:
Ok. Fair enough